Clean up Intel Compiler Warnings

wesbland opened the pull request:
A bunch of warning fixes that come up with the Intel compiler along with a few bug fixes that the compiler caught as warnings.

wesbland commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

wesbland commented:
Test comment

pavanbalaji commented:
I've rebased the branch on to the latest `mpich/master` and signed off on some of the commits.

wesbland commented:
Responding to comments from @pavanbalaji 

>In `Restrict values of errflag to the enumeration`, the change seems awkward. In some cases, errflag is set to a class and in some cases it's set to a code (i.e., `MPIR_ERR_PROC_FAILED`. This needs to be made consistent to always point to the error class.

Actually, `MPIR_ERR_PROC_FAILED` and `MPIR_ERR_OTHER` aren't error classes. They're enum values defined here: https://github.com/pmodels/mpich/blob/1f7b75a797d94f601311f1c50bfddd829e8b59ff/src/include/mpir_misc.h#L23-L27

The point of `errflag` was to track whether a process failed or anything else happened. The specific error class is tracked via the status object.

>In `romio: Avoid uninitilized warnings`, we now have double initialization of `io_thread`, e.g., it is assigned to pthread_self for a second time in GPFS. This usage should be cleaned up.

👍 Done

>In `mpi: Guard fn_fail with error checking ifdef` and` mpi_t: Guard fn_fail with error checking ifdef`, the comment does not seem to match the change. Also, it is unclear why this change is needed at all. `mpi_errno` is not being assigned at all before the check.

The reason those were added is that `mpi_errno` is only set during the `#ifdef HAVE_ERROR_CHECKING` sections. If we put all of the `fn_fail` macros inside the same macros, we'd also have to do the same with `fn_exit` (because it's only called from `fn_fail`). I'm not really sure what the best solution is. The tradeoff is an extra check vs. a whole bunch of `#ifdef`s. I'd started down the second route and changed my mind and apparently forgot to change the commit message to reflect that.

>In `mpid: Guard fn_fail with error checking ifdef`, it's better if the function flag usage is made consistent with the other functions, i.e., have both `fn_exit` and `fn_fail`. In an error-free execution flowing through to `fn_fail` seems weird.

I agree, though we'll have the same tradeoff as above. Maybe I could just add another macro that combines all of the `#ifdef`s so it doesn't look so gross.

>In `romio: Guard fn_fail with error checking ifdef`, the commit message doesn't match the change.

Yep. I think this was for the same reason as above. I've fixed that.

wesbland commented:
I've addressed all of the concerns from @pavanbalaji. There's another warning that's come up that might get in here, but let's move forward without it for now.

test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
I've approved this PR.  I'll squash a few commits, signoff and merge to master once the tests finish.
